src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -1655,7 +1714,10 @@ private void ValidateTransition(WriterState newState)
 
                     break;
                 case WriterState.ResourceSet:
-                    if (newState != WriterState.Resource)
+                    if (newState != WriterState.Resource &&
+                        (this.CurrentScope.ResourceType == null ||
+                            (this.CurrentScope.ResourceType.TypeKind != EdmTypeKind.Untyped ||
+                                (newState != WriterState.Primitive && newState != WriterState.ResourceSet))))

```

Comment:
my poor brain is reeling from this. can you provide a comment for what drives the complex state check and what it amounts to?